generators/typescript.go

```diff
@@ -0,0 +1,287 @@
+			signers: [],
+		})
+
+		if (res.stack.length === 0) {
+			throw new Error(res.exception ?? 'unrecognized response')
+		}
+		{{- if ne .ReturnType "void"}}
+		
+		return this.config.parser.parseRpcResponse(res.stack[0], { type: '{{ .ReturnTypeABI }}' })
+		{{- end}}
+	}
+{{- end -}}
+{{- define "TESTMETHOD" }}
+	{{- if eq .ReturnTypeABI "InteropInterface" }}
+	{{- template "ITERATORGENERATORMETHOD" . -}}
+	{{ else }}
+	{{- template "TESTINVOKEMETHOD" . -}}
+	{{- end -}}
+{{- end -}}
+{{- define "EVENTLISTENER" }}
+	async confirm{{ upperFirst .Name }}Event(txId: string): Promise<void>{
+		if (!this.config.eventListener) throw new Error('EventListener not provided')
+
+		const txResult = await this.config.eventListener.waitForApplicationLog(txId)
+		this.config.eventListener.confirmTransaction(
+			txResult, {contract: this.config.scriptHash, eventname: '{{ .Name }}'}
+		)
+	}
+
+	listen{{ upperFirst .Name }}Event(callback: Neo3EventListenerCallback): void{
+		if (!this.config.eventListener) throw new Error('EventListener not provided')
+		
+		this.config.eventListener.addEventListener(this.config.scriptHash, '{{ .Name }}', callback)
+	}
+
+	remove{{ upperFirst .Name }}EventListener(callback: Neo3EventListenerCallback): void{
+		if (!this.config.eventListener) throw new Error('EventListener not provided')
+		
+		this.config.eventListener.removeEventListener(this.config.scriptHash, '{{ .Name }}', callback)
+	}
+{{- end -}}
+import { Neo3EventListener, Neo3EventListenerCallback, Neo3Invoker, Neo3Parser } from "@cityofzion/neon-dappkit-types"
+import { typeChecker } from "@cityofzion/neon-dappkit"
+import * as Invocation from './api'
+
+export type SmartContractConfig = {
+  scriptHash: string;
+  invoker: Neo3Invoker;
+  parser?: Neo3Parser;
+  eventListener?: Neo3EventListener | null;
+}
+
+export class {{ .ContractName }}{
+  static SCRIPT_HASH = '{{ .Hash }}'
+
+  private config: Required<SmartContractConfig>
+
+	constructor(configOptions: SmartContractConfig) {
+		this.config = { 
+			...configOptions, 
+			parser: configOptions.parser ?? require("@cityofzion/neon-dappkit").NeonParser,
+			eventListener: configOptions.eventListener ?? null
+		}
+	}
+
+{{- range $e := .Events}}
+{{ template "EVENTLISTENER" $e -}}
+{{end}}
+{{- range $m := .Methods}}
+{{if .Safe -}}
+{{ template "TESTMETHOD" $m -}}
+{{- else -}}
+{{ template "INVOKEMETHOD" $m }}
+{{ template "TESTMETHOD" $m -}}
+{{end -}}
+{{end}}
+}
+`
+
+const typescriptSrcIndexTmpl = `export * from './{{ .ContractName }}'
+export * from './api'`
+
+func GenerateTypeScriptSDK(cfg *GenerateCfg) error {
+	cfg.MethodNameConverter = strcase.ToLowerCamel
+	cfg.ParamTypeConverter = scTypeToTypeScript
+	ctr, err := templateFromManifest(cfg)
+	if err != nil {
+		return fmt.Errorf("failed to parse manifest into contract template: %v", err)
+	}
+
+	sdkDir := cfg.SdkDestination + ctr.ContractName
```

Comment:
is there a naming convention in TS that says if folders should be all lower case? I looked at `neon-js` and `neon-dappkit` and all folders seem to be lower case but right now it generates the folder with upper case first letter and the rest whatever was defined as the name in the contract.

Reply:
While there is no official TS naming convention, we decided to change it to kebab-case to simulate a js package.

Reply:
all good 👍